eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec/Invalid.hs

```diff
@@ -2,228 +2,347 @@
+      -- Would be an Iso if microlens supported Iso's
+      redeemersL :: Lens' (Redeemers era) (Map.Map (PlutusPurpose AsIx era) (Data era, ExUnits))
+      redeemersL = lens unRedeemers (const $ Redeemers @era)
+      -- PlutusPurpose serialization wasn't fixed until Conway
+      withPPRTFailures =
+        if eraProtVerLow @era < natVersion @9
+          then withCborRoundTripFailures
+          else id
+
+  forM_ (eraLanguages @era) $ \lang ->
+    withSLanguage lang $ \slang ->
+      describe (show lang) $ do
+        let redeemerSameAsDatumHash = hashPlutusScript $ redeemerSameAsDatum slang
+            alwaysSucceedsWithDatumHash = hashPlutusScript $ alwaysSucceedsWithDatum slang
+            alwaysSucceedsNoDatumHash = hashPlutusScript $ alwaysSucceedsNoDatum slang
+
+        it "MissingRedeemers" $ do
+          let scriptHash = redeemerSameAsDatumHash
+          txIn <- produceScript scriptHash
+          let missingRedeemer = mkSpendingPurpose $ AsItem txIn
+          let tx = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+          withPostFixup (fixupResetAddrWits . (witsTxL . rdmrsTxWitsL .~ mempty)) $
+            submitFailingTx
+              tx
+              [ injectFailure $
+                  MissingRedeemers [(missingRedeemer, scriptHash)]
+              , injectFailure $
+                  CollectErrors [NoRedeemer missingRedeemer]
+              ]
+
+        it "MissingRequiredDatums" $ do
+          txIn <- produceScript redeemerSameAsDatumHash
+          let tx = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+          let missingDatum = hashData @era (Data (P.I 3))
+          withPostFixup (fixupResetAddrWits . (witsTxL . datsTxWitsL .~ mempty)) $
+            submitFailingTx
+              tx
+              [injectFailure $ MissingRequiredDatums [missingDatum] []]
+
+        it "NotAllowedSupplementalDatums" $ do
+          txIn <- produceScript redeemerSameAsDatumHash
+          let extraDatumHash = hashData @era (Data (P.I 30))
+          let extraDatum = Data (P.I 30)
+          let tx =
+                mkBasicTx mkBasicTxBody
+                  & bodyTxL . inputsTxBodyL .~ [txIn]
+                  & witsTxL . datsTxWitsL .~ TxDats (Map.singleton extraDatumHash extraDatum)
+          submitFailingTx
+            tx
+            [injectFailure $ NotAllowedSupplementalDatums [extraDatumHash] []]
+
+        describe "PPViewHashesDontMatch" $ do
+          let
+            test hash = do
+              txIn <- produceScript redeemerSameAsDatumHash
+              tx <- fixupTx $ mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+              wrongIntegrityHashTx <-
+                resetAddrWits $ tx & bodyTxL . scriptIntegrityHashTxBodyL .~ hash
+              withNoFixup $
+                submitFailingTx
+                  wrongIntegrityHashTx
+                  [injectFailure $ PPViewHashesDontMatch hash (tx ^. bodyTxL . scriptIntegrityHashTxBodyL)]
```

Comment:
That doesn't quite work, unfortunately, because the second hash in the `PPViewHashesDontMatch` failure needs to be the correct one, and the `tx` we get from `submitFailingTxM` is the one that results from running `messupScriptIntegrityHash`. The tree-diffs are:

```
             (PPViewHashesDontMatch
               (SJust
                 (SafeHash
                   "d5b21fb06ca5253d9ac86a619c56659565a96914bf3ba9b16d3ee87e601067c4"))
               (SJust
                 (SafeHash
                   -"7070e33696da2429af91a928147703371e2947f85394039569833f8c0d23a696"
                   +"d5b21fb06ca5253d9ac86a619c56659565a96914bf3ba9b16d3ee87e601067c4")))]
```

and

```
             (PPViewHashesDontMatch
               SNothing
               -(SJust
                 (SafeHash
                   "7070e33696da2429af91a928147703371e2947f85394039569833f8c0d23a696"))
               +SNothing)]
```

_ie_ we're saying the expected hash is the one we've messed it up with.

I'll play with this some more tomorrow.

Reply:
I don't think it can be done with `submitFailingTxM`, unfortunately. Instead, I've added a commit that makes the code a bit clearer by changing the variable names a bit.

Reply:
It was worht a try :slightly_smiling_face:

Reply:
For sure!